### PR TITLE
[PHP] Add Swoole

### PR DESCRIPTION
**Pull request by stdex**
Before submitting your PR, please review the following checklist :

## If you are adding a framework

+ [x] Please check configuration files
   + [x] `neph.yml`, for [neph](http://tbrand.github.io/neph) job processing
   + [x] `FRAMEWORKS.yml`, for **frameworks** list
   + [x] `travis.yml`, for CI
+ [x] A `Dockerfile` exists ?
+ [x] All tests passes ?
~~~
export FRAMEWORK=<MY_FRAMEWORK>
shards install
shards build --release
bin/neph ${FRAMEWORK}
crystal spec
~~~

```
| php (7.3) | [swoole](http://swoole.co.uk) (master) | 7.25 ms | 6.67 ms | 15.13 ms | 25.31 ms | 72.50 ms | 5861.33 | 
| php (7.3) | [swoole](http://swoole.co.uk) (master) | 147949.67 | 375.20 MB |
```

### CODE REVIEW

**Comment by waghanza** [comment]
Hi @stdex,

Thanks for your contribution. 

As I understand, this `PR` is more to add **fast-route** than **swoole**.

**Swoole** will be an `async` replacement for **nginx/php-fpm**. 

I think the best option for us is to replace all php implementations here, with **swoole**

**Comment by stdex** [comment]
@waghanza 

1. As you can see:
>ENTRYPOINT php /usr/src/app/public/index.php start
This mean, that we start swoole server (through php-cli, without php-fpm and nginx). This is the main difference with other php benchmarks (where uses php-fpm+nginx).
2. Swoole does not include a router, so chosen as the most lightweight router (common practice, e.g.: https://github.com/easy-swoole/http/blob/master/composer.json#L20)
3. This is not framework, this is custom solution without additional overhead.

If you want I can change benchmark name to swoole_fast-route

**Comment by waghanza** [comment]
@stdex I understand what you're saying,  but my concern is about :
+ `swoole` is not a **framework**
+ `swoole` could not be compared to, for example, `symfony`

BTW, I think naming this **swoole_fast-route** is acceptable

File: php/swoole_fast-route/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM twosee/swoole-coroutine:latest
```
Comment (waghanza) [inline php/swoole_fast-route/Dockerfile]: I think `zaherg/php-swoole`  is more accurate

File: FRAMEWORKS.yml [hunk 2]
```diff
@@ -172,6 +172,10 @@ php:
     website: framework.zend.com
     version: "3.1"
     language: 7.3
+  swoole_fast-route:
+    website: swoole.co.uk
+    version: "master"
```
Comment (waghanza) [inline FRAMEWORKS.yml]: please fix the version

File: php/swoole_fast-route/Dockerfile (hunk 1 above)
Comment (waghanza) [reply php/swoole_fast-route/Dockerfile]: at least it use `pecl` -> allowing us to not use `swoole`  **master** but the last stable (4.3 as for now)

File: php/swoole_fast-route/composer.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+    "require": {
+        "nikic/fast-route": "^1.3"
```
Comment (waghanza) [inline php/swoole_fast-route/composer.json]: better using `~` instead of `^`

File: php/swoole_fast-route/public/index.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    $r->addRoute('POST', '/user', 'post_user');
+}, [
+    'cacheFile' => __DIR__ . '/route.cache',
+]);
+
+$host = '0.0.0.0';
+$port = 3000;
+
+$server = new Server($host, $port, SWOOLE_BASE);
+$server->set(['worker_num' => swoole_cpu_num() * 2]);
```
Comment (waghanza) [inline php/swoole_fast-route/public/index.php]: why 2 workers per core ?

**Comment by waghanza** [comment]
May I ask for your permission for editing you `PR` ?

I think I've found a way to use `swoole` alongside `php/fpm`

**Comment by stdex** [comment]
@waghanza
yes, you can edit, if it improve solution performance

File: php/swoole_fast-route/Dockerfile (hunk 1 above)
Comment (stdex) [reply php/swoole_fast-route/Dockerfile]: It's a bad idea. 
In my test this leads to performance reduce. (It's maybe alpine vs debian or pecl vs compile or etc problems).
On my hw:
1st:
```
Requests / s | Throughput |
138286.67 | 350.80 MB |
```
2nd:
```
Requests / s | Throughput |
81526.33 | 206.29 MB |
```

File: php/swoole_fast-route/public/index.php (hunk 4 above)
Comment (stdex) [reply php/swoole_fast-route/public/index.php]: value from official benchmark: https://github.com/swoole/swoole-src/blob/master/benchmark/benchmark.php#L5

File: php/swoole_fast-route/Dockerfile (hunk 1 above)
Comment (waghanza) [reply php/swoole_fast-route/Dockerfile]: sure, I understand the performance gap, but using **master** leads to results fluctuating, I mean we can show any version

**Comment by waghanza** [comment]
@stdex This project is about having representative results for any **web-frameworks**, including `router` only **web-frameworks**

btw, after https://github.com/the-benchmarker/web-frameworks/pull/1336, I think it is great feature to represent `php` **web-frameworks** with several variants :
+ `swoole`
+ `php-pm`
+ `php-fpm` (since php built-in)

As for example, we have a performance gap (but also stability gap) between native (`php-fpm`) and async (`swoole`)

| | Average | 50th percentile | 90th percentile | 99th percentile | 99.9th percentile | Standard deviation |
|----------------------------|----------------:|----------------:|----------------:|----------------:|----------------:|----------------:|
| swoole (4.3) | 18.32 ms | 14.47 ms | 40.87 ms | 71.97 ms | 145.94 ms | 16361.00 |
| php-fpm (7.3) | 183.03 ms | 0.82 ms | 389.20 ms | 3437.41 ms | 7278.98 ms | 617079.67 |
| swoole (master) | 9.81 ms | 7.36 ms | 23.41 ms | 38.41 ms | 78.83 ms | 9476.67 |

PS : In here there is still **docker** noise -> results are not very accurate until https://github.com/the-benchmarker/web-frameworks/pull/632

File: php/swoole_fast-route/public/index.php (hunk 4 above)
Comment (waghanza) [reply php/swoole_fast-route/public/index.php]: ok

**Comment by waghanza** [comment]
@stdex I will close your `PR` since I have open https://github.com/the-benchmarker/web-frameworks/issues/1481

As this project code base is in complete rewrite, I'll do this later

Thanks for your contribution

**Pull request closed without merging**